llpc/patch/llpcPatchIntrinsicSimplify.cpp

```diff
@@ -120,20 +123,30 @@ bool PatchIntrinsicSimplify::runOnFunction(
             {
                 continue;
             }
-            Value* const pSimplifiedValue = Simplify(*pIntrinsicCall);
 
-            // We did not simplify the intrinsic call.
-            if (pSimplifiedValue == nullptr)
+            // Record intrinsic only if it can be simplified.
+            if (CanSimplify(*pIntrinsicCall))
             {
-                continue;
+                candidateCalls.push_back(pIntrinsicCall);
             }
+        }
+    }
 
-            changed = true;
+    // Process all intrinsics which can be simplified.
+    for (IntrinsicInst* const pIntrinsicCall : candidateCalls) {
+        Value* const pSimplifiedValue = Simplify(*pIntrinsicCall);
 
-            pIntrinsicCall->replaceAllUsesWith(pSimplifiedValue);
-            pIntrinsicCall->eraseFromParent();
-            m_pScalarEvolution->eraseValueFromMap(pIntrinsicCall);
+        // We did not simplify the intrinsic call.
+        if (pSimplifiedValue == nullptr)
+        {
+            continue;
         }
+
+        changed = true;
+
+        pIntrinsicCall->replaceAllUsesWith(pSimplifiedValue);
+        pIntrinsicCall->eraseFromParent();
+        m_pScalarEvolution->eraseValueFromMap(pIntrinsicCall);
```

Comment:
`erase` deletes the value and invalidates the `pInstrinsicCall` pointer. This seems like a use of a invalidated pointer, which is an undefined behavior in C++. Should SCEV be taken care of before erasing the value, or alternatively, replace the `erase` with `remove`, and then `delete` manually after cleaning up SCEV?